blocks/bricks/sequence_generators.py

```diff
@@ -136,6 +137,34 @@ def _push_allocation_config(self):
 
     @application
     def cost(self, application_call, outputs, mask=None, **kwargs):
+        """Returns sum of generation costs for output sequences.
+
+        Parameters
+        ----------
+        outputs : :class:`~tensor.TensorVariable`
+            The 3(2) dimensional tensor containing output sequences.
+            The dimension 0 must stand for time, the dimension 1 for the
+            position on the batch.
+        mask : :class:`~tensor.TensorVariable`
+            The binary matrix identifying fake outputs.
+
+        Notes
+        -----
+        The contexts are expected as keyword arguments.
+
+        """
+        # Compute the sum of costs
+        cost = self.cost_matrix(outputs, mask=mask, **kwargs).sum()
```

Comment:
On Mon, Mar 16, 2015 at 1:53 PM, Bart van Merriënboer <
notifications@github.com> wrote:

> In blocks/bricks/sequence_generators.py
> <https://github.com/bartvm/blocks/pull/472#discussion_r26510001>:
>
> > +        Parameters
> > +        ----------
> > +        outputs : :class:`~tensor.TensorVariable`
> > +            The 3(2) dimensional tensor containing output sequences.
> > +            The dimension 0 must stand for time, the dimension 1 for the
> > +            position on the batch.
> > +        mask : :class:`~tensor.TensorVariable`
> > +            The binary matrix identifying fake outputs.
> > +
> > +        Notes
> > +        -----
> > +        The contexts are expected as keyword arguments.
> > +
> > +        """
> > +        # Compute the sum of costs
> > +        cost = self.cost_matrix(outputs, mask=mask, **kwargs).sum()
>
> @rizar <https://github.com/rizar>'s proposal sounds good. I guess the
> case of widely variable batch sizes isn't common enough to warrant the pain
> of having to change your learning rate every time you change batch size.
> @kyunghyuncho <https://github.com/kyunghyuncho>, do you normalize the
> cost by dividing by the batch size in GroundHog too?
>
Yes. The cost should be an average over a minibatch but a sum over
timesteps.


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bartvm/blocks/pull/472/files#r26510001>.
>